Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs

```diff
@@ -1,110 +1,408 @@
 {
-    private float _updateTimer = 0.0f;
+    [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+    [Dependency] private readonly SharedMapSystem _sharedMapSystem = default!;
+
+    private List<EntityUid> _trackedDevices = new();
+    private bool _powerNetAbnormalities = false;
+    private const float RoguePowerConsumerThreshold = 100000;
+    private float _updateTimer = 1.0f;
     private const float UpdateTime = 1.0f;
 
-    [Dependency] private readonly NodeContainerSystem _nodeContainer = default!;
-    [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<PowerMonitoringConsoleComponent, RequestPowerMonitoringUpdateMessage>(OnUpdateRequestReceived);
+        SubscribeLocalEvent<ExpandPvsEvent>(OnExpandPvsEvent);
+        SubscribeLocalEvent<GameRuleStartedEvent>(OnPowerGridCheckStarted);
+        SubscribeLocalEvent<GameRuleEndedEvent>(OnPowerGridCheckEnded);
+    }
 
     public override void Update(float frameTime)
     {
         _updateTimer += frameTime;
+
         if (_updateTimer >= UpdateTime)
         {
             _updateTimer -= UpdateTime;
 
-            var query = EntityQueryEnumerator<PowerMonitoringConsoleComponent>();
-            while (query.MoveNext(out var uid, out var component))
+            // On update, update our list of anchored power monitoring devices
+            _trackedDevices.Clear();
+
+            var query = AllEntityQuery<PowerMonitoringDeviceComponent, TransformComponent>();
+            while (query.MoveNext(out var ent, out var _, out var xform))
             {
-                UpdateUIState(uid, component);
+                if (xform.Anchored)
+                    _trackedDevices.Add(ent);
             }
         }
     }
 
-    public void UpdateUIState(EntityUid target, PowerMonitoringConsoleComponent? pmcComp = null, NodeContainerComponent? ncComp = null)
+    // Sends the list of tracked power monitoring devices to all player sessions with one or more power monitoring consoles open
+    // This expansion of PVS is needed so that the sprites for these device are available to the the player UI
+    // Out-of-range devices will be automatically removed from the player PVS when the UI closes
+    private void OnExpandPvsEvent(ref ExpandPvsEvent ev)
     {
-        if (!Resolve(target, ref pmcComp))
+        var query = AllEntityQuery<PowerMonitoringConsoleComponent>();
+        while (query.MoveNext(out var uid, out var _))
+        {
+            if (_userInterfaceSystem.SessionHasOpenUi(uid, PowerMonitoringConsoleUiKey.Key, ev.Session))
+            {
+                if (ev.Entities == null)
+                    ev.Entities = new List<EntityUid>();
+
+                foreach (var ent in _trackedDevices)
+                    ev.Entities.Add(ent);
+
+                break;
+            }
+        }
+    }
+
```

Comment:
This looks very expensive as you're gonna be at least doubling or tripling entities in range realistically and I really think you should do something else.

Reply:
As per our discussion on Discord, it's only like to be a 20-30% increase in the center of the station. Substantial, yes, but not terrible. I can reduce this down to a 10-15% increase

Reply:
I've optimized this further. In busy areas on the station, the increase in the player PVS will be less than 10%